### ISSUE TITLE
Mesh_3 bug: The surface of c3t3 has holes

### ISSUE THREAD

**Issue by lrineau** [CGAL/cgal#1554]
With the .cpp and the data shipped in https://gist.github.com/5e71e899c85b949487b0316e1f6e064b, one can see that the surfacic part of the c3t3 has holes.

The issue is that the circumcenters are not computed exactly, and results of queries to the domain oracle are not coherent.

I think (to be confirmed) that bug was not a bug before the commit fbf824c7113f6c9f3f9faf95cd8d458a596cc162 that merged 'Mesh_3-parallel-cjamin' into 'master': the circumcenters of cells were cached into cells, and thus computed only once.

CC @janetournois @cjamin

**Comment by cjamin** [comment]
It is a bit far for me to remember. What changed in the code regarding these cached circumcenters?

**Comment by lrineau** [comment]
Before that merge, the computation of the dual was done by a call to `tr.dual(..)`, and that function was forwarding to `c->circumcenter()`, allowing the cell type to have a cache.

**Comment by lrineau** [comment]
I have tried this patch with hope but it was not sufficient:
```patch
diff --git a/Mesh_3/include/CGAL/Mesh_3/Refine_facets_3.h b/Mesh_3/include/CGAL/Mesh_3/Refine_facets_3.h
index d23b771..1770cf2 100644
--- a/Mesh_3/include/CGAL/Mesh_3/Refine_facets_3.h
+++ b/Mesh_3/include/CGAL/Mesh_3/Refine_facets_3.h
@@ -1577,16 +1577,8 @@ dual_segment(const Facet & facet, Bare_point& p, Bare_point& q) const
   int i = facet.second;
   Cell_handle n = c->neighbor(i);
   CGAL_assertion( ! r_tr_.is_infinite(c) && ! r_tr_.is_infinite(n) );
-  p = Gt().construct_weighted_circumcenter_3_object()(
-      c->vertex(0)->point(),
-      c->vertex(1)->point(),
-      c->vertex(2)->point(),
-      c->vertex(3)->point());
-  q = Gt().construct_weighted_circumcenter_3_object()(
-      n->vertex(0)->point(),
-      n->vertex(1)->point(),
-      n->vertex(2)->point(),
-      n->vertex(3)->point());
+  p = r_tr_.dual(c);
+  q = r_tr_.dual(n);
 }
 
 template<class Tr, class Cr, class MD, class C3T3_, class Ct, class C_>
@@ -1640,11 +1632,7 @@ dual_ray(const Facet & facet, Ray_3& ray) const
     ( Gt().construct_plane_3_object()(p,q,r),
       Gt().construct_weighted_circumcenter_3_object()(p,q,r) );
 
- ray = Gt().construct_ray_3_object()(Gt().construct_weighted_circumcenter_3_object()(
-        n->vertex(0)->point(),
-        n->vertex(1)->point(),
-        n->vertex(2)->point(),
-        n->vertex(3)->point()), l);
+  ray = Gt().construct_ray_3_object()(r_tr_.dual(n), l);
 }
 
 template<class Tr, class Cr, class MD, class C3T3_, class Ct, class C_>
```

**Comment by cjamin** [comment]
In my e-mails, I could find several discussion from 2012 between you, Mariette and Monique about the cached circumcenters. Seems that for regular triangulations, they were computed but not used...
You mention a branch you had at this time: Triangulation_3-regular_cell_base_with_circumcenter-GF
Not sure what happened to it...

**Comment by lrineau** [comment]
It has been merged into CGAL-4.5 (see [release notes]),  by e97506adb7ae27fa4aa10726813629db76262957. In the same release CGAL-4.5 the parallel version for Dt_3, Rt_3 and Mesh_3 was published.

[release notes]: http://www.cgal.org/2014/10/10/cgal-45/

**Comment by lrineau** [comment]
Fixed by #1554.
